### PR TITLE
[RFC] allow event processing in getchar()

### PR DESCRIPTION
**Pull request by bfredl**
this is consistent with vim, and is necessary for plugins that implement
their own input modes using `getchar()` and still want to do async
event processing while waiting for input, such as Denite.

simple test using timers:

```
" 10
let g:x = 0

func! Demo1(timer)
    let g:x += 1
    if has("nvim")
        call nvim_buf_set_lines(0,0,1,v:true,['" '.g:x])
    else
        py3 import vim
        py3 vim.current.buffer[0] = '" {}'.format(vim.vars["x"])
    endif
    redraw
endfunc

let g:timer = timer_start(500, function("Demo1"), {"repeat": 10})

let g:char = getchar()
```

explicit `redraw` is needed in both vim and neovim, I think this is correct, so that async processing can still avoid unnecessary redraws if it wants. `redraw` should be used when implementing a user interface like Denite and, say, async job output causes the list of canditates to change.

### CODE REVIEW

**Comment by bfredl** [comment]
oops, the test "timers are paused when event processing is disabled" was broken to start with (cannot fail), the repeat count should have been -1

**Comment by justinmk** [comment]
> explicit redraw is needed in both vim and neovim, I think this is correct

Agreed.  https://github.com/neovim/neovim/issues/3416#issuecomment-145282126 mentions

> we still have to execute redraw checks to ensure asynchronous calls are visible by the user.

but... why? Perhaps we should disable that totally (in a separate PR, of course) and see what happens.

File: src/nvim/eval.c [hunk 1]
```diff
@@ -9501,24 +9501,35 @@ static void f_getchar(typval_T *argvars, typval_T *rettv, FunPtr fptr)
...
+    ui_cursor_goto(msg_row, msg_col);
+
+    if (argvars[0].v_type == VAR_UNKNOWN) {
+      // getchar(): blocking wait.
+      if (!(char_avail() || using_script() || input_available())) {
+        input_enable_events();
+        (void)os_inchar(NULL, 0, -1, 0);
+        input_disable_events();
+        if (!queue_empty(main_loop.events)) {
+          queue_process_events(main_loop.events);
```
Comment (justinmk) [inline src/nvim/eval.c]: `event/queue.h` was renamed to `multiqueue.h`, so this is now `multiqueue_process_events`, `multiqueue_empty`, etc.

**Comment by justinmk** [comment]
LGTM, is this ready?

**Comment by bfredl** [comment]
> but... why? Perhaps we should disable that totally (in a separate PR, of course) and see what happens.

IMO there is a difference between "ordinary" editing modes like normal and insert where the user should be able to expect to always see the current buffer state (or at least, redraws should not be delayed indefinitely) , which I think was @tarruda:s main point, and "modal" ("more modal than usual") states like command-line, `getchar()` and "press ENTER", in this cases redraws should only be shown if it is really relevant (like incsearch/incsub, and an async handler intentionaly redrawing as it is relevant to the input)

I also think that it should not only be possible to implement a Unite/Denite-like interface using `input()`/command-line-mode, it also should be relatively easy to do so. That might require #4419 ,  some cmdline-mode autocommands (I hope to open PR soon), and `setcmdline()` function would be nice while we're at it. (yet another project I would work on when/if I have more free time...)

File: src/nvim/eval.c (hunk 1 above)
Comment (bfredl) [inline src/nvim/eval.c]: Thanks for the headsup, updated.

**Comment by justinmk** [comment]
Looks like there might be an issue with the test on the travis macOS environment.

**Comment by bfredl** [comment]
Increased the tolerance a bit, if this truly breaks it should still be very likely that the test fails.

**Pull request merged**